p2p/src/lib.rs

```diff
@@ -54,6 +75,69 @@ pub trait Sender: Clone + Debug + Send + Sync + 'static {
     ) -> impl Future<Output = Result<Vec<Self::PublicKey>, Self::Error>> + Send;
 }
 
+/// Interface for constructing a [`CheckedSender`] from a set of [`Recipients`],
+/// filtering out any that are currently rate-limited.
+pub trait LimitedSender: Clone + Send + Sync + 'static {
+    /// Public key type used to identify recipients.
+    type PublicKey: PublicKey;
+
+    /// The type of [`GClock`] used for rate-limiting.
+    type Clock: GClock + Send + 'static;
```

Comment:
Note to reviewer - kept the `GClock` associated type to remove the need to propagate `GClock<Instant = SystemTime>` throughout the codebase. All of our backing implementations are `SystemTime` anyways - we should be able to just add that bound at the callsites where we need it (i.e. the resolver.)